### PR TITLE
feat(optimistic-oracle-proposer): Add dispute logic

### PR DESCRIPTION
**Pull request by nicholaspai**
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

**Motivation**

Part of #2411

Adds dispute logic to Optimistic Oracle proposer bot.

**Summary**

- Added event parameters to `OptimisticOracle.sol` to simplify bot logic. See comments for more details.
- Implemented `sendDisputes` in the `packages/OptimisticOracle: proposer.js` bot which checks proposed prices for `undisputed proposals` and sends disputes where `disputePrice != proposePrice`

Future work:
- Add a disputePrice buffer configuration variable to the OO bot which allows `disputePrice` and `proposePrice` to diverge beyond a certain precision, (e.g. 5 decimals of precision), without having to send a dispute.
- Add settlePrice feature

**Details**

- Adding event parameters to `OptimisticOracle.sol` should be harmless but its important to note that this would change Contract bytecode from that in the latest OpenZeppelin audit.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (nicholaspai) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:97]: This event parameter allows the bot to directly approve any collateral currencies from `ProposePrice` events, which would be used by the bot if for some reason the collateral currency from a proposal was not already approved when it was pre-proposal.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (nicholaspai) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:106]: This event parameter allows the bot to more easily report the proposedPrice for a dispute transaction, which is useful for the bot owner to know.

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.js
Comment (nicholaspai) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.js:79]: Now handled in `proposer.js` code

File: packages/optimistic-oracle/src/proposer.js
Comment (nicholaspai) [inline packages/optimistic-oracle/src/proposer.js:341]: This is where [this new contract event](https://github.com/UMAprotocol/protocol/pull/2546/files#r577015775) is used.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (mrice32) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:97]: Just playing devil's advocate: could we just approve if we decide to dispute and look up the full request in that case? Or does that cause other problems? I only say that because each of these adds about 200 gas to each log message.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (mrice32) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:106]: Similar to the above, aren't disputes pretty rare? Would it be too painful to just directly query the request in question if there's a dispute?

**Review by mrice32 — COMMENTED** [review]
Just wanted to give quick feedback on the contract changes: while they don't seem like they would cause any security issues, just curious _how much_ extra trouble not having them creates for the bot? Both seem like a lookup would only be required in the case of a dispute, which (at least for now) we expect to be fairly rare, so the extra lookup wouldn't be too bad, at least on our request count. But would be curious to hear how much worse they would be for our bot logic.

**Comment by nicholaspai** [comment]
> Just wanted to give quick feedback on the contract changes: while they don't seem like they would cause any security issues, just curious _how much_ extra trouble not having them creates for the bot? Both seem like a lookup would only be required in the case of a dispute, which (at least for now) we expect to be fairly rare, so the extra lookup wouldn't be too bad, at least on our request count. But would be curious to hear how much worse they would be for our bot logic.

We've already made changes to the `OptimisticOracle.sol` contract bytecode since the OZ audit, see [this PR](https://github.com/UMAprotocol/protocol/pull/2395) for example. I don't see how this PR adds additional uncertainty. WDYT?

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:97]: also if we moved over to a DSProxy model we would be able to do the approval in the same time as the dispute logic and would not need do any pre-approval logic.

However, not really any downside (other than gas) to adding thins IMHO.

File: packages/optimistic-oracle/src/proposer.js
Comment (chrismaree) [inline packages/optimistic-oracle/src/proposer.js:218]:
```suggestion
      // If proposal price is not equal to the dispute price, then prepare dispute
```

File: packages/optimistic-oracle/src/proposer.js
Comment (chrismaree) [inline packages/optimistic-oracle/src/proposer.js:219]: this is a great point and something we need to consider quite closely when it comes to funding rates. ideally a funding rate within a small margin of error cant be disputed as it'll become basically impossible to create a third party implementation that does not constantly get disputed due to small variants. There needs to be some window of tolerance for marginally wrong proposals that _dont_ get disputed. WDYT?

File: packages/optimistic-oracle/src/proposer.js
Comment (chrismaree) [inline packages/optimistic-oracle/src/proposer.js:269]: do you not want to use `ynatm`?

**Review by chrismaree — APPROVED** [review]
LGTM! nice and clean code, well commented and tested. good work

File: packages/optimistic-oracle/src/proposer.js
Comment (nicholaspai) [reply packages/optimistic-oracle/src/proposer.js:218]: I'll open up a subsequent PR on this I have one idea where we basically check precision up to X decimals of precision. For example, 1.2345 and 1.23456 are "equal" up to 4 decimals of precision

File: packages/optimistic-oracle/src/proposer.js
Comment (nicholaspai) [reply packages/optimistic-oracle/src/proposer.js:268]: I think we should deprecate using `ynatm` since we don't use it in the `disputer`, and its easier to not have to maintain an external repo

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (nicholaspai) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:97]: The reason why I like blanket approving all collateral during the `update` method is so that we can better isolate failures during the `sendDisputes` and `sendProposals` method to the actual propose/dispute logic. If for some reason the approvals fail, then we'd want these failures to happen as early as possible, not right before a dispute needs to be sent.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (nicholaspai) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:106]: Disputes are rare, but this is also why I see little harm adding a param to this event

Comment (mrice32) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:106]: That's fair, the gas impact will probably be low.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol
Comment (mrice32) [reply packages/core/contracts/oracle/implementation/OptimisticOracle.sol:97]: My very weak opinion is that it doesn't seem worth the extra 400 gas (200 for the read and 200 for the extra log field) for a bot optimization -- however, I'm cool either way. LGTM!

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**
